debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/processor/AbstractLogMinerEventProcessor.java

```diff
@@ -0,0 +1,650 @@
+                        row.getTableId(),
+                        row.getRowId(),
+                        row.getRsId(),
+                        row.getChangeTime()));
+    }
+
+    /**
+     * Handle processing a LogMinerEventRow for a {@code INSERT}, {@code UPDATE}, or {@code DELETE} event.
+     *
+     * @param row the result set row
+     * @throws SQLException if a database exception occurs
+     * @throws InterruptedException if the dispatch of an event is interrupted
+     */
+    protected void handleDataEvent(LogMinerEventRow row) throws SQLException, InterruptedException {
+        if (row.getRedoSql() == null) {
+            return;
+        }
+
+        LOGGER.trace("DML: {}", row);
+        LOGGER.trace("\t{}", row.getRedoSql());
+
+        counters.dmlCount++;
+        switch (row.getEventType()) {
+            case INSERT:
+                counters.insertCount++;
+                break;
+            case UPDATE:
+                counters.updateCount++;
+                break;
+            case DELETE:
+                counters.deleteCount++;
+                break;
+        }
+
+        final TableId tableId = row.getTableId();
+        Table table = getSchema().tableFor(tableId);
+        if (table == null) {
+            if (!getConfig().getTableFilters().dataCollectionFilter().isIncluded(tableId)) {
+                return;
+            }
+            table = dispatchSchemaChangeEventAndGetTableForNewCapturedTable(tableId, offsetContext, dispatcher);
+        }
+
+        if (row.isRollbackFlag()) {
+            // There is a use case where a constraint violation will result in a DML event being
+            // written to the redo log subsequently followed by another DML event that is marked
+            // with a rollback flag to indicate that the prior event should be omitted. In this
+            // use case, the transaction can still be committed, so we need to manually rollback
+            // the previous DML event when this use case occurs.
+            final Transaction transaction = getCache().get(row.getTransactionId());
+            if (transaction == null) {
+                LOGGER.warn("Cannot undo change '{}' since transaction was not found.", row);
+            }
+            else {
+                transaction.getEvents().removeIf(o -> {
+                    if (o.getRowId().equals(row.getRowId())) {
+                        LOGGER.trace("Undo applied for '{}'.", row);
+                        return true;
+                    }
+                    return false;
+                });
+            }
+            return;
+        }
+
+        final LogMinerDmlEntry dmlEntry = parseDmlStatement(row.getRedoSql(), table, row.getTransactionId());
+        dmlEntry.setObjectName(row.getTableName());
+        dmlEntry.setObjectOwner(row.getTablespaceName());
+
+        addToTransaction(row.getTransactionId(),
+                row,
+                () -> new DmlEvent(row.getEventType(),
```

Comment:
I'll add a second constructor that accepts a `LogMinerEventRow` for this use case but I think we need to keep the other so that the adapter code doesn't need to create an event row before creating the events when read from disk.